File: frontend/app/react-native/src/api/MateuApiClient.ts
export interface RunActionParams {
  route: string;
  consumedRoute: string;
  actionId: string;
  serverSideType: string | null;
  initiatorComponentId: string;
  componentState: Record<string, unknown>;
  appState: Record<string, unknown>;
  parameters: Record<string, unknown>;
}

export class MateuApiClient {
  private readonly baseUrl: string;
  private readonly sessionId: string;

  constructor(baseUrl: string, sessionId: string) {
    this.baseUrl = baseUrl;
    this.sessionId = sessionId;
  }

  async runAction(params: RunActionParams): Promise<unknown> {
    const routeStripped = params.route.startsWith('/') ? params.route.slice(1) : params.route;
    const urlSegment = routeStripped === '' ? '_no_route' : routeStripped;
    const bodyRoute = routeStripped === '' ? '' : '/' + routeStripped;

    const body: Record<string, unknown> = {
      route: bodyRoute,
      consumedRoute: params.consumedRoute ?? '',
      actionId: params.actionId ?? '',
      serverSideType: params.serverSideType,
      initiatorComponentId: params.initiatorComponentId,
      componentState: params.componentState ?? {},
      appState: params.appState ?? {},
      parameters: params.parameters ?? {},
    };

    // Drop null values to match web frontend behaviour
    for (const key of Object.keys(body)) {
      if (body[key] === null || body[key] === undefined) {
        delete body[key];
      }
    }

    const url = `${this.baseUrl}/mateu/v3/sync/${urlSegment}`;
    console.log('[Mateu] --> POST', url);
    console.log('[Mateu]     body:', JSON.stringify(body).slice(0, 500));

    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
        'X-Session-Id': this.sessionId,
      },
      body: JSON.stringify(body),
    });

    const text = await response.text();
    console.log('[Mateu] <--', response.status, text.slice(0, 500));

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${text}`);
    }

    return JSON.parse(text);
  }

  async initialLoad(route: string, appState: Record<string, unknown>): Promise<unknown> {
    return this.runAction({
      route,
      consumedRoute: '_empty',
      actionId: '',
      serverSideType: null,
      initiatorComponentId: 'ux_main',
      componentState: {},
      appState,
      parameters: {},
    });
  }

  async navigate(
    route: string,
    consumedRoute: string,
    serverSideType: string,
    appState: Record<string, unknown>,
  ): Promise<unknown> {
    return this.runAction({
      route,
      consumedRoute: consumedRoute || '_empty',
      actionId: '',
      serverSideType: serverSideType || null,
      initiatorComponentId: 'ux_main',
      componentState: {},
      appState,
      parameters: {},
    });
  }

  async runFormAction(
    route: string,
    consumedRoute: string,
    actionId: string,
    serverSideType: string,
    componentId: string,
    componentState: Record<string, unknown>,
    appState: Record<string, unknown>,
  ): Promise<unknown> {
    return this.runAction({
      route,
      consumedRoute: consumedRoute || '_empty',
      actionId,
      serverSideType: serverSideType || null,
      initiatorComponentId: componentId,
      componentState,
      appState,
      parameters: {},
    });
  }
}

File: frontend/app/compose/src/commonMain/kotlin/io/mateu/compose/api/MateuApiClient.kt
package io.mateu.compose.api

import io.ktor.client.HttpClient
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/** Each platform provides its Ktor engine (desktop: CIO; later Android: OkHttp, iOS: Darwin, web: Js). */
expect fun createHttpClient(): HttpClient

/**
 * Multiplatform HTTP client for the Mateu sync endpoint (Ktor + kotlinx.serialization).
 * Wire contract unchanged: `POST /{baseUrl}/mateu/v3/sync/{route}` → `UIIncrementDto` JSON.
 */
class MateuApiClient(
    private val baseUrl: String,
    private val sessionId: String,
    private val http: HttpClient = createHttpClient(),
) {
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    suspend fun runAction(
        route: String?,
        consumedRoute: String?,
        actionId: String?,
        serverSideType: String?,
        initiatorComponentId: String?,
        componentState: Map<String, Any?>,
        appState: Map<String, Any?>,
        parameters: Map<String, Any?>,
    ): JsonElement {
        val routeStripped = when {
            route != null && route.startsWith("/") -> route.substring(1)
            route != null -> route
            else -> ""
        }
        val urlSegment = if (routeStripped.isEmpty()) "_no_route" else routeStripped
        val bodyRoute = if (routeStripped.isEmpty()) "" else "/$routeStripped"

        val body = buildJsonObject {
            put("route", bodyRoute)
            put("consumedRoute", consumedRoute ?: "")
            put("actionId", actionId ?: "")
            // Omit null values so the JSON matches the web frontend's behaviour
            if (serverSideType != null) put("serverSideType", serverSideType)
            if (initiatorComponentId != null) put("initiatorComponentId", initiatorComponentId)
            put("componentState", anyToJson(componentState))
            put("appState", anyToJson(appState))
            put("parameters", anyToJson(parameters))
        }

        val url = "$baseUrl/mateu/v3/sync/$urlSegment"
        val payload = body.toString()
        println("[Mateu] --> POST $url")
        println("[Mateu]     body: $payload")

        val response = http.post(url) {
            contentType(ContentType.Application.Json)
            header("Accept", "application/json")
            header("X-Session-Id", sessionId)
            setBody(payload)
        }
        val text = response.bodyAsText()
        println("[Mateu] <-- ${response.status.value}")
        println(
            "[Mateu]     response (first 3000 chars): " +
                if (text.length > 3000) text.substring(0, 3000) + "..." else text,
        )
        if (response.status.value >= 400) {
            throw RuntimeException("HTTP ${response.status.value}: $text")
        }
        return json.parseToJsonElement(text)
    }

    suspend fun initialLoad(route: String?, appState: Map<String, Any?>): JsonElement =
        runAction(route, "_empty", "", null, "ux_main", emptyMap(), appState, emptyMap())

    suspend fun navigate(
        route: String?,
        consumedRoute: String?,
        serverSideType: String?,
        appState: Map<String, Any?>,
    ): JsonElement =
        runAction(route, consumedRoute ?: "_empty", "", serverSideType, "ux_main", emptyMap(), appState, emptyMap())
}

/** Converts the loosely-typed values held in componentState/parameters into JSON. */
private fun anyToJson(value: Any?): JsonElement = when (value) {
    null -> JsonNull
    is JsonElement -> value
    is String -> JsonPrimitive(value)
    is Boolean -> JsonPrimitive(value)
    is Int -> JsonPrimitive(value)
    is Long -> JsonPrimitive(value)
    is Double -> JsonPrimitive(value)
    is Float -> JsonPrimitive(value)
    is Map<*, *> -> buildJsonObject { value.forEach { (k, v) -> put(k.toString(), anyToJson(v)) } }
    is List<*> -> buildJsonArray { value.forEach { add(anyToJson(it)) } }
    else -> JsonPrimitive(value.toString())
}
